File: src/entity/player/player.lua
function create (obj)
	obj:setMaxHitpoints (25)
    obj:setCurrentHitpoints(25)
	obj:setAttack (15)
	obj:setDefence (12)
	obj:setSpeed (14)
	obj:setExperience (0)
end

function battle (player, enemy)
    -- speed
    p_speed = player : getSpeed()
    e_speed = enemy : getSpeed()

    -- attack
    p_attack = player : getAttack()
    e_attack = enemy : getAttack()

    -- defence
    p_defence = player : getDefence()
    e_defence = enemy : getDefence()

    -- names
    p_name = player : getName()
    e_name = enemy : getName()

    print("===== Battle Starting! =====")

    if(p_speed > e_speed) then
        -- player attacks
        print(p_name, " attacks for ", damageCheck(p_attack, e_defence), " damage!")
        calc = enemy : getHitpoints() - damageCheck(p_attack, e_defence)
        enemy : setHitpoints(calc)

        if(enemy : isAlive()) then
            print(e_name, " has ", enemy : getHitpoints(), " hitpoints remaining!")
            print(e_name, " attacks for ", damageCheck(e_attack, p_defence), " damage!")
            calc = player : getCurrentHitpoints() - damageCheck(e_attack, p_defence)
            player : setCurrentHitpoints(calc)

            if(player : isAlive()) then
                print(p_name, " has ", player:getCurrentHitpoints(), " hitpoints remaining!")
            else
                player : defeated()
            end
        else
            print(e_name, " is defeated!")
            enemy : defeated()
        end
    else
        -- enemy attacks
        print(e_name, " attacks for ", damageCheck(e_attack, p_defence), " damage!")

        if(player : isAlive()) then
            print(p_name, " has ", enemy:getHitpoints(), " hitpoints remaining!")
            print(p_name, " attacks for ", damageCheck(p_attack, e_defence), " damage!")

            calc = enemy:getHitpoints() - damageCheck(p_attack, e_defence)
            enemy : setHitpoints(calc)

            if(enemy : isAlive()) then
                print(e_name, " has ", enemy:getHitpoints(), " hitpoints remaining!")
            else
                enemy : defeated()
            end
        else
            player : defeated()
        end
    end -- end combat

    print("===== Battle Complete =====")

end

-- Evaluates how much damage was dealt for this round of combat
function damageCheck(attack, defence)
	if(defence >= attack) then
		return 1
	else
		return (attack - defence)
	end
end

File: src/entity/player/Player.java
package entity.player;

import assets.item.Drop;
import assets.item.InventoryItem;
import util.LoadScript;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Represents the main character in the game who you use to battle with
 * Created by Trevor on 6/11/2016.
 */
public class Player {
    protected String name;
    protected int attack, defence, speed, maxHitpoints, currentHitpoints, experience;
    protected HashMap<String, InventoryItem> inventory;
    private LoadScript scriptLoader;

    public Player(String name){
        this.name = name;
        this.inventory = new HashMap<>();

        scriptLoader = new LoadScript("src/entity/player/player.lua");
        scriptLoader.runScriptFunction("create", this, 1, 0);
    }

    public LoadScript getScriptLoader(){
        return scriptLoader;
    }

    // Prints to the console that the game is over and the player is dead
    public void defeated(){
        System.out.println("The hero has fallen. Game Over.");
    }

    /**
     *  Determine if the player is alive based on hitpoints
     */
    public boolean isAlive(){
        if(currentHitpoints > 0)
            return true;

        return false;
    }

    /**
     * Adds a new Item to the player's inventory from a monster drop.
     * Accordingly updates amount if the item already exists
     */
    public void addDrop(Drop drop) {
        if (inventory.containsKey(drop.getItem().getName())) {
            int oldAmount = inventory.get(drop.getItem().getName()).getAmount();
            inventory.put(drop.getItem().getName(), new InventoryItem(drop.getItem(), oldAmount + drop.getAmount()));
        } else {
            inventory.put(drop.getItem().getName(), new InventoryItem(drop.getItem(), drop.getAmount()));
        }
    }

    /**
     * Prints out the players current stats to the console
     */
    public void displayStats(){
        System.out.println("========= Player Stats =========");
        System.out.println("Name: " + name);
        System.out.println("Health: " + currentHitpoints + "/" + maxHitpoints);
        System.out.println("Attack: " + attack);
        System.out.println("Defense: " + defence);
        System.out.println("Speed: " + speed);
        System.out.println("Experience: " + experience);
        System.out.println("================================");
    }

    /**
     * Prints out the players inventory to the console
     */
    public void displayInventory(){
        System.out.println("Player's Inventory:");
        Iterator it = inventory.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry pair = (Map.Entry)it.next();
            InventoryItem entry = (InventoryItem) pair.getValue(); // convert the key to the actual item to read its name

            System.out.println(pair.getKey() + " : " + entry.getAmount());
        }
    }

    /**
     * Accessors and Mutators
     */
    public String getName() { return name; }
    public int getAttack() {
        return attack;
    }
    public void setAttack(int attack) {
        this.attack = attack;
    }
    public int getDefence() {
        return defence;
    }
    public void setDefence(int defence) {
        this.defence = defence;
    }
    public int getSpeed() {
        return speed;
    }
    public void setSpeed(int speed) {
        this.speed = speed;
    }
    public int getCurrentHitpoints() {
        return currentHitpoints;
    }
    public void setCurrentHitpoints(int hitpoints) {
        this.currentHitpoints = hitpoints;
    }
    public int getMaxHitpoints() {return maxHitpoints;}
    public void setMaxHitpoints(int maxHitpoints) { this.maxHitpoints = maxHitpoints; }
    public int getExperience() {
        return experience;
    }
    public void setExperience(int experience) {
        this.experience = experience;
    }
    public HashMap<String, InventoryItem> getInventory() {
        return inventory;
    }
    public void setInventory(HashMap<String, InventoryItem> inventory) { this.inventory = inventory; }
}
